core/src/main/java/org/apache/brooklyn/core/location/AbstractMachineLocation.java

```diff
@@ -0,0 +1,139 @@
+
+    public static final ConfigKey<Boolean> DETECT_MACHINE_DETAILS = ConfigKeys.newBooleanConfigKey("detectMachineDetails",
+            "Attempt to detect machine details automatically.", true);
+
+    protected static final MachineDetails UNKNOWN_MACHINE_DETAILS = new BasicMachineDetails(
+            new BasicHardwareDetails(-1, -1),
+            new BasicOsDetails("UNKNOWN", "UNKNOWN", "UNKNOWN")
+    );
+
+    private volatile MachineDetails machineDetails;
+    private final Object machineDetailsLock = new Object();
+
+    private transient WithMutexes mutexSupport;
+    private transient final Object mutexSupportCreationLock = new Object();
+
+
+    public AbstractMachineLocation() {
+        this(MutableMap.of());
+    }
+
+    public AbstractMachineLocation(Map<?,?> properties) {
+        super(properties);
+    }
+
+    /**
+     * Returns the machine details only if they are already loaded, or available directly as
+     * config.
+     */
+    protected Optional<MachineDetails> getOptionalMachineDetails() {
+        MachineDetails result = machineDetails != null ? machineDetails : config().get(MACHINE_DETAILS);
+        return Optional.fromNullable(result);
+    }
+
+    @Override
+    public MachineDetails getMachineDetails() {
+        synchronized (machineDetailsLock) {
+            if (machineDetails == null) {
+                machineDetails = getConfig(MACHINE_DETAILS);
+            }
+            if (machineDetails == null) {
+                boolean detectionEnabled = getConfig(DETECT_MACHINE_DETAILS);
+                if (!detectionEnabled || !isManaged()) {
+                    return UNKNOWN_MACHINE_DETAILS;
+                }
+                machineDetails = detectMachineDetails();
+            }
+            return machineDetails;
+        }
+    }
+
+    @Override
+    public OsDetails getOsDetails() {
+        return getMachineDetails().getOsDetails();
+    }
+
+    protected abstract MachineDetails detectMachineDetails();
+
+    public WithMutexes mutexes() {
+        synchronized (mutexSupportCreationLock) {
+            // create on demand so that it is not null after serialization
+            if (mutexSupport == null) {
+                mutexSupport = new MutexSupport();
+            }
+            return mutexSupport;
+        }
+    }
+
+    /** @deprecated Mutex-related methods are now accessible via {@link #mutexes()} */
+    @Override
+    @Deprecated
+    public void acquireMutex(String mutexId, String description) throws RuntimeInterruptedException {
+        try {
+            mutexes().acquireMutex(mutexId, description);
+        } catch (InterruptedException ie) {
+            throw new RuntimeInterruptedException("Interrupted waiting for mutex: " + mutexId, ie);
+        }
+    }
+
+    /** @deprecated Mutex-related methods are now accessible via {@link #mutexes()} */
+    @Override
+    @Deprecated
+    public boolean tryAcquireMutex(String mutexId, String description) {
```

Comment:
Why have these on the new abstract super class? If we want people to use `mutexes().tryAcquireMutex()` instead, then we should leave these methods only on `SshMachineLocation`, and also `AbstractMachineLocation` should not implement `WithMutexes`. It seems wrong to add deprecated methods to the new super-class.

Reply:
Agree.